### PR TITLE
Add semver 2 versioning in dependabot common

### PR DESCRIPTION
**Pull request by amazimbe**
### What are you trying to accomplish?
Implement semver 2 (https://semver.org/spec/v2.0.0.html) as the standard that dependabot follows for ecosystems that have no version standards; docker for example.

We have numerous errors and issues due to lack of clarity on what a valid dependency version should be and how to consistently compare versions within an ecosystem. Gem::Version has been inadequate so far and we think semver 2 will be an improvement especially for ecosystems without a well defined standard.

Malformed version errors and issues related to dependabot opening PRs to update a dependency to an unacceptable version, for example, a prerelease or beta release.

### Anything you want to highlight for special attention from reviewers?
This one is the first in a series of PRs. Subsequent PRs will focus on improving ecosystem specific versioning by implementing the standard for the ecosystem where one exists; implementing deviations from semver 2 where applicable and improving test coverage which is below 50% in most cases.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: common/spec/dependabot/sem_version_spec.rb [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+      versions << described_class.new("1.0.0-alpha.1")
+      versions << described_class.new("1.0.0-alpha.1.beta.gamma")
+      versions << described_class.new("1.0.0-alpha.beta")
+      versions << described_class.new("1.0.0-alpha.beta.1")
+      versions << described_class.new("1.0.0-beta")
+      versions << described_class.new("1.0.0-beta.2")
+      versions << described_class.new("1.0.0-beta.11")
+      versions << described_class.new("1.0.0-rc.1")
+      versions << described_class.new("1.0.0")
+      expect(versions.shuffle.sort).to eq(versions)
```
Comment (abdulapopoola) [inline common/spec/dependabot/sem_version_spec.rb:166]: Can we have more tests to cover the other methods in the base class please? Also, what's the current test coverage?

Comment (amazimbe) [reply common/spec/dependabot/sem_version_spec.rb:166]: 96.55% It's got the highest coverage of all the files in dependabot common.

File: common/lib/dependabot/sem_version.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "sem_version"
+require "sorbet-runtime"
+
+module Dependabot
+  class SemVersion < Dependabot::Version
```
Comment (jakecoffman) [inline common/lib/dependabot/sem_version.rb:8]: Not sure why we would inherit from Dependabot::Version if the intent is to make it obsolete.

File: Gemfile [hunk 3]
```diff
@@ -39,6 +39,7 @@ deps_shared_with_common = %w(
   rubocop-performance
   rubocop-rspec
   rubocop-sorbet
+  sem_version
```
Comment (jakecoffman) [inline Gemfile:42]: Is this a library we want to depend on? It seems there's an issue from 2019 which has never been acknowledged: https://github.com/canton7/sem_version/issues/2

File: common/spec/dependabot/sem_version_spec.rb [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    context "with a release" do
+      let(:version_string) { "1.0.0" }
+
+      it { is_expected.to be(false) }
+    end
+
+    context "with a + separated alphanumeric build identifier" do
+      let(:version_string) { "1.0.0+build1" }
+
+      it { is_expected.to be(false) }
```
Comment (abdulapopoola) [inline common/spec/dependabot/sem_version_spec.rb:142]: is there a reason this is false (i.e. not correct?)

**Review by abdulapopoola — APPROVED** [review]

**Review by jakecoffman — COMMENTED** [review]
Can you write more tests with known valid/invalid values to make sure this implementation supports valid SemVer 2.0? For example, some folks have listed known values here: https://github.com/semver/semver/issues/833

File: Gemfile (hunk 3 above)
Comment (amazimbe) [reply Gemfile:42]: > Is this a library we want to depend on? It seems there's an issue from 2019 which has never been acknowledged: [canton7/sem_version#2](https://github.com/canton7/sem_version/issues/2)

That definitely is worrying. I'm happy to remove this dependency and implement semver 2 from scratch as it is quite small.

File: common/lib/dependabot/sem_version.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
+# typed: strong
+# frozen_string_literal: true
+
+require "sorbet-runtime"
+
+# See https://semver.org/spec/v2.0.0.html for semver 2 details
+#
+module Dependabot
+  class SemVersion
```
Comment (abdulapopoola) [inline common/lib/dependabot/sem_version.rb:9]: maybe rename and call this SemVer2 for clarity?

Comment (amazimbe) [reply common/lib/dependabot/sem_version.rb:9]: I can do that.

File: common/lib/dependabot/sem_version.rb [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    end
+
+    sig { params(other: ::Dependabot::SemVersion).returns(T::Boolean) }
+    def eql?(other)
+      other.is_a?(self.class) && to_s == other.to_s
+    end
+
+    sig { params(other: ::Dependabot::SemVersion).returns(Integer) }
+    def <=>(other)
+      maj = major.to_i <=> other.major.to_i
```
Comment (abdulapopoola) [inline common/lib/dependabot/sem_version.rb:72]: maybe rename from `maj` since this is actually a comparison?

**Comment by amazimbe** [comment]
I'm very happy with the code and branch test coverage both of which are now on 💯 %

**Review by jakecoffman — APPROVED** [review]

**Pull request merged**